src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1744,6 +1745,87 @@ class Simplex_tree {
     }
   }
 
+ public:
+  /** \brief Serialize the Simplex tree - Flatten it in a vector of char */
+  std::vector<char> serialize() {
+    std::vector<char> buffer{};
+    rec_serialize(&root_, buffer);
```

Comment:
From a quick test (`buffer.reserve(num_simplices()*32);` where 32 is not the right number), it seems that I was wrong about the cost of calling num_simplices() compared to vector reallocations.
Even after `reserve`, `vec.insert(vec.end(),...)` has a lot of overhead compared to just writing the number and incrementing the pointer, although that depends a lot on how I compile (serialize takes twice as long with clang -03 than with gcc -O3 and an extra attribute or profile feedback).
Could you experiment a bit with a version where we allocate the right size (using num_simplices) at the beginning and then write numbers and increment the pointer ourselves instead of relying on `vector::insert`?

Reply:
I did it on [04e5a85](https://github.com/GUDHI/gudhi-devel/pull/810/commits/04e5a8541cc11e79d5c69aa7e0d7ad965c9a88d2) and benchmarks are better with this version:
# C++

## Rips de 250 points sur le 3-tore, expansion en dimension 2
`vector::insert` | `num_simplices`
---|---
Rips creation: 0.003s             | Rips creation: 0.001s
Simplex tree creation: 0.043s     | Simplex tree creation: 0.039s
Serialize Simplex_tree: 0.055s    | Serialize Simplex_tree: 0.022s
Deserialize Simplex_tree: 0.034s  | Deserialize Simplex_tree: 0.033s

## Rips de 250 points sur le 3-tore, expansion en dimension 3
`vector::insert` | `num_simplices`
---|---
Rips creation: 0.002s             | Rips creation: 0.002s
Simplex tree creation: 2.062s     | Simplex tree creation: 2.072s
Serialize Simplex_tree: 3.488s    | Serialize Simplex_tree: 1.078s
Deserialize Simplex_tree: 1.979s  | Deserialize Simplex_tree: 1.84s

## Rips de 8000 points sur le 3-tore, expansion en dimension 1
`vector::insert` | `num_simplices`
---|---
Rips creation: 1.072s             | Rips creation: 1.056s
Simplex tree creation: 0.881s     | Simplex tree creation: 0.882s
Serialize Simplex_tree: 0.596s    | Serialize Simplex_tree: 0.258s
Deserialize Simplex_tree: 0.522s  | Deserialize Simplex_tree: 0.359s

## Rips de 16000 points sur le 3-tore, expansion en dimension 1
`vector::insert` | `num_simplices`
---|---
Rips creation: 4.22s              | Rips creation: 4.286s
Simplex tree creation: 3.78s      | Simplex tree creation: 3.744s
Serialize Simplex_tree: 2.379s    | Serialize Simplex_tree: 0.985s
Deserialize Simplex_tree: 2.267s  | Deserialize Simplex_tree: 1.38s

# Python
## Rips de 250 points sur le 3-tore, expansion en dimension 2
`vector::insert` | `num_simplices`
---|---
Rips creation: 0.002s            | Rips creation: 0.002s 
Simplex tree creation: 0.057s    | Simplex tree creation: 0.057s 
Serialize Simplex_tree: 0.101s   | Serialize Simplex_tree: 0.064s
Deserialize Simplex_tree: 0.077s | Deserialize Simplex_tree: 0.063s 
File size: 40Mo

## Rips de 250 points sur le 3-tore, expansion en dimension 3
`vector::insert` | `num_simplices`
---|---
Rips creation: 0.002s            | Rips creation: 0.002s
Simplex tree creation: 2.312s    | Simplex tree creation: 2.459s
Serialize Simplex_tree: 5.952s   | Serialize Simplex_tree: 3.588s
Deserialize Simplex_tree: 4.380s | Deserialize Simplex_tree: 3.278s
File size: 2.5Go

## Rips de 8000 points sur le 3-tore, expansion en dimension 1
`vector::insert` | `num_simplices`
---|---
Rips creation: 1.208s            | Rips creation: 1.216s
Simplex tree creation: 1.003s    | Simplex tree creation: 1.077s
Serialize Simplex_tree: 1.083s   | Serialize Simplex_tree: 0.783s
Deserialize Simplex_tree: 0.993s | Deserialize Simplex_tree: 0.686s
File size: 489Mo

## Rips de 16000 points sur le 3-tore, expansion en dimension 1
`vector::insert` | `num_simplices`
---|---
Rips creation: 4.871s            | Rips creation: 5.045s
Simplex tree creation: 4.460s    | Simplex tree creation: 4.509s
Serialize Simplex_tree: 4.308s   | Serialize Simplex_tree: 2.862s
Deserialize Simplex_tree: 4.024s | Deserialize Simplex_tree: 2.556s
File size: 2Go

Reply:
> I did it on [04e5a85](https://github.com/GUDHI/gudhi-devel/pull/810/commits/04e5a8541cc11e79d5c69aa7e0d7ad965c9a88d2) and benchmarks are better with this version:

Thanks, that's indeed an impressive difference.
It is disappointing that compilers don't do almost as well with the other version :disappointed: